src/Shared/Internal/RendererReEntrantManager.cs

```diff
@@ -0,0 +1,120 @@
+using System;
+#if NET35
+using System.Web;
+#else
+using System.Threading;
+#endif
+
+namespace NLog.Web.Internal
+{
+    /// <summary>
+    /// Manages if a LayoutRenderer can be called recursively.
+    /// Especially used by AspNetSessionValueLayoutRenderer
+    /// 
+    /// Since NET 35 does not support AsyncLocal or even ThreadLocal
+    /// a different technique must be used for that platform
+    /// </summary>
+    internal class RendererReEntrantManager : IDisposable
+    {
+#if NET35
+        private static readonly object ReEntrantLock = new object();
+
+        internal RendererReEntrantManager(HttpContextBase context)
+        {
+            httpContext = context;
+        }
+
+        private readonly HttpContextBase httpContext;
+
+        private bool ReadLock()
+        {
+            return httpContext?.Items?.Contains(ReEntrantLock) == true;
+        }
+
+        private void Lock()
+        {
+            httpContext.Items[ReEntrantLock] = bool.TrueString;
+        }
+
+        private void Unlock()
+        {
+            httpContext?.Items?.Remove(ReEntrantLock);
+        }
+#else
+        // Manage access to the session re-entrancy, at least above .NET 3.5
+        private static readonly AsyncLocal<bool> ReEntrantLock = new AsyncLocal<bool>();
+
+        internal RendererReEntrantManager()
+        {
+
+        }
+
+        private bool ReadLock()
+        {
+            return ReEntrantLock.Value;
+        }
+
+        private void Lock()
+        {
+            ReEntrantLock.Value = true;
+        }
+
+        private void Unlock()
+        {
+            ReEntrantLock.Value = false;
+        }
+#endif
+
+        // Need to track if we were successful in the entry
+        // If we were not, we should not unlock in the dispose code
+        private bool entrySuccess;
+
+
+        internal bool TryGetLock()
```

Comment:
Could `TryGetLock()` become private and called by the constructor, and then make new property `bool IsLockAcquired => entrySuccess` that is used instead from outside code.

Reply:
Done as requested